{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eef3c5e4-8eaf-4df7-96f8-ec372a1e6b76",
   "metadata": {},
   "source": [
    "# Python pour biochimistes: BioPython pour interroger des bases de données distantes\n",
    "\n",
    "## Introduction\n",
    "\n",
    "Dans un cahier précédent, nous avons développé la logique nécessaire pour accéder des ressources distantes: ouverture d'un flux de données (*handle*), construction d'un URL contenant les éléments de la requete, extraction de l'info, etc. C'est pas si complexe mais ça reste toujours un exercice long et pas toujours plaisant car ça demande de connaitre la façon de prendre contact avec la ressource distante. Pourquoi ne pas utiliser une approche orientée-objet pour faciliter notre vie: encapsuler les actions à prendre sous forme de méthodes avec un usage prévisible?\n",
    "\n",
    "BioPython permet de faire ce type d'action et dans l'exercice qui suit, nous utiliserons `Bio.Entrez` pour intérroger directement Entrez. Noter que plusieurs autres sources de données sont aussi accessibles via BioPython; voir la documentation pour les détails. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47ce8ed6-4876-4485-b780-a830db99e830",
   "metadata": {},
   "source": [
    "## Protocole pour obtenir des infos sur notre gène préféré\n",
    "\n",
    "### Étape 1 - Quelles sont les bases de données accessibles via `Bio.Entrez` \n",
    "\n",
    "Accéder à une ressource distante demande un certain décorum... Par exemple, tel que déjà fait dans un exemple précédent, il faut généralement s'identifier. De plus, il faut souvent limiter ces demandes de requêtes pour ne pas inonder le serveur distant et risquer un bannissement. Ici, passer par BioPython est un avantage car Bio.Entrez force une limite de 3 requêtes/seconde."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "72622026-49b3-43ef-a1c1-1859765a70f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Quelles sont les bases de donnees accessibles?\n",
      "pubmed\n",
      "protein\n",
      "nuccore\n",
      "ipg\n",
      "nucleotide\n",
      "structure\n",
      "genome\n",
      "annotinfo\n",
      "assembly\n",
      "bioproject\n",
      "biosample\n",
      "blastdbinfo\n",
      "books\n",
      "cdd\n",
      "clinvar\n",
      "gap\n",
      "gapplus\n",
      "grasp\n",
      "dbvar\n",
      "gene\n",
      "gds\n",
      "geoprofiles\n",
      "medgen\n",
      "mesh\n",
      "nlmcatalog\n",
      "omim\n",
      "orgtrack\n",
      "pmc\n",
      "popset\n",
      "proteinclusters\n",
      "pcassay\n",
      "protfam\n",
      "pccompound\n",
      "pcsubstance\n",
      "seqannot\n",
      "snp\n",
      "sra\n",
      "taxonomy\n",
      "biocollections\n",
      "gtr\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Primo, importons les classes nécessaires\n",
    "#\n",
    "import xml.dom.minidom\n",
    "import xmltodict\n",
    "from Bio import Entrez, SeqIO\n",
    "#\n",
    "# C'est toujours une bonne pratique de nous faire connaitre\n",
    "#\n",
    "Entrez.email = \"sylvain.foisy@umontreal.ca\"\n",
    "#\n",
    "# Création automatique du flux de données\n",
    "#\n",
    "handle = Entrez.einfo()\n",
    "#\n",
    "# Interception du résultat de la méthode einfo()\n",
    "#\n",
    "result = handle.read()\n",
    "#\n",
    "# On n'utilisera plus le format XML directement\n",
    "# Si vous voulez voir la structure du XML,\n",
    "# retirez les commentaires\n",
    "#\n",
    "# Impression du résultat et observation du format\n",
    "#\n",
    "#demoformat = xml.dom.minidom.parseString(result)\n",
    "#print(demoformat.toprettyxml())\n",
    "#\n",
    "# Conversion du format XML en dictionnaire Python\n",
    "#\n",
    "aLib = xmltodict.parse(result)\n",
    "#\n",
    "# Quelles sont les bases de données accessibles?\n",
    "#\n",
    "print(\"Quelles sont les bases de donnees accessibles?\")\n",
    "for i in aLib[\"eInfoResult\"][\"DbList\"][\"DbName\"]:\n",
    "  #\n",
    "  # On pourrait nettre i dans une liste pour réutilisation \n",
    "  # dans votre script\n",
    "  #\n",
    "  print(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "139757f6-b684-46fd-9ff5-0db4cf9d6252",
   "metadata": {},
   "source": [
    "### Étape 2 - Ok, on a une vue d'ensemble... Et ensuite? Recherche via Entrez.esearch\n",
    "\n",
    "On voit que la liste des infos potentielles est vaste :-) Comment s'en tirer? Disons que nos voulons chercher des infos sur notre gène préféré: la cortactine humaine, alias CTTN."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9442dc3c-4d57-4f04-9f31-4e0d8021f167",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'eSearchResult': {'Count': '1', 'RetMax': '1', 'RetStart': '0', 'IdList': {'Id': '2017'}, 'TranslationSet': None, 'TranslationStack': {'TermSet': [{'Term': 'CTTN[Gene]', 'Field': 'Gene', 'Count': '485', 'Explode': 'N'}, {'Term': '9606[Taxid]', 'Field': 'Taxid', 'Count': '359812', 'Explode': 'N'}], 'OP': 'AND'}, 'QueryTranslation': 'CTTN[Gene] AND 9606[Taxid]'}}\n",
      "Quel est le geneID de CTTN chez H. sapiens?\n",
      "2017\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Cherchons dans NCBI Gene pour CTTN chez H sapiens, alias 9606 dans le jargon Taxid\n",
    "#\n",
    "aSearch = Entrez.esearch(db=\"gene\", term=\"CTTN[Gene] AND 9606[Taxid]\")\n",
    "result = aSearch.read()\n",
    "newLib = xmltodict.parse(result)\n",
    "print(newLib)\n",
    "#\n",
    "# On voit que ça ne retourne qu'un seul résultat\n",
    "# avec le geneId 2017\n",
    "#\n",
    "print(\"Quel est le geneID de CTTN chez H. sapiens?\")\n",
    "print(newLib[\"eSearchResult\"][\"IdList\"][\"Id\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6204d577-48ab-4ec1-99b1-639cd002cc2a",
   "metadata": {},
   "source": [
    "### Étape 3 - Récupération des infos via Entrez.efetch\n",
    "\n",
    "Comme NCBI Gene est le dépositaire centralisé des infos sur les gènes recensés, récupérons dans un premier temps les info sur notre gène :-)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d51fd02-1ac4-4aea-b3df-5852cd102e50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTTN\n",
      "cortactin\n",
      "This gene is overexpressed in breast cancer and squamous cell carcinomas of the head and neck. The encoded protein is localized in the cytoplasm and in areas of the cell-substratum contacts. This gene has two roles: (1) regulating the interactions between components of adherens-type junctions and (2) organizing the cytoskeleton and cell adhesion structures of epithelia and carcinoma cells. During apoptosis, the encoded protein is degraded in a caspase-dependent manner. The aberrant regulation of this gene contributes to tumor cell invasion and metastasis. Three splice variants that encode different isoforms have been identified for this gene. [provided by RefSeq, May 2010]\n",
      "CTTN cortactin[Homo sapiens]\n",
      "Gene ID: 2017, updated on 17-Jun-2024\n",
      "\n",
      "\n",
      "Reference GRCh38.p14 Primary Assembly NC_000011.10  from: 70398529 to: 70436575\n",
      "mRNA transcript variant 1 NM_005231.4, 18 exons,  total annotated spliced exon length: 3249\n",
      "protein isoform a NP_005222.2 (CCDS41680.1), 16 coding  exons,  annotated AA length: 550\n",
      "\n",
      "Exon table for  mRNA  NM_005231.4 and protein NP_005222.2\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t543\n",
      "70420400-70420510\t\t70420400-70420510\t\t21872-21982\t\t21872-21982\t\t111\t\t111\t\t959\n",
      "70421470-70421580\t\t70421470-70421580\t\t22942-23052\t\t22942-23052\t\t111\t\t111\t\t1359\n",
      "70422940-70422995\t\t70422940-70422995\t\t24412-24467\t\t24412-24467\t\t56\t\t56\t\t2336\n",
      "70425332-70425401\t\t70425332-70425401\t\t26804-26873\t\t26804-26873\t\t70\t\t70\t\t3649\n",
      "70429051-70429199\t\t70429051-70429199\t\t30523-30671\t\t30523-30671\t\t149\t\t149\t\t1991\n",
      "70431191-70431280\t\t70431191-70431280\t\t32663-32752\t\t32663-32752\t\t90\t\t90\t\t1820\n",
      "70433101-70433278\t\t70433101-70433278\t\t34573-34750\t\t34573-34750\t\t178\t\t178\t\t368\n",
      "70433647-70433718\t\t70433647-70433718\t\t35119-35190\t\t35119-35190\t\t72\t\t72\t\t1307\n",
      "70435026-70436575\t\t70435026-70435162\t\t36498-38047\t\t36498-36634\t\t1550\t\t137\n",
      "\n",
      "mRNA transcript variant 2 NM_138565.3, 17 exons,  total annotated spliced exon length: 3138\n",
      "protein isoform b NP_612632.1 (CCDS8197.1), 15 coding  exons,  annotated AA length: 513\n",
      "\n",
      "Exon table for  mRNA  NM_138565.3 and protein NP_612632.1\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t543\n",
      "70420400-70420510\t\t70420400-70420510\t\t21872-21982\t\t21872-21982\t\t111\t\t111\t\t2429\n",
      "70422940-70422995\t\t70422940-70422995\t\t24412-24467\t\t24412-24467\t\t56\t\t56\t\t2336\n",
      "70425332-70425401\t\t70425332-70425401\t\t26804-26873\t\t26804-26873\t\t70\t\t70\t\t3649\n",
      "70429051-70429199\t\t70429051-70429199\t\t30523-30671\t\t30523-30671\t\t149\t\t149\t\t1991\n",
      "70431191-70431280\t\t70431191-70431280\t\t32663-32752\t\t32663-32752\t\t90\t\t90\t\t1820\n",
      "70433101-70433278\t\t70433101-70433278\t\t34573-34750\t\t34573-34750\t\t178\t\t178\t\t368\n",
      "70433647-70433718\t\t70433647-70433718\t\t35119-35190\t\t35119-35190\t\t72\t\t72\t\t1307\n",
      "70435026-70436575\t\t70435026-70435162\t\t36498-38047\t\t36498-36634\t\t1550\t\t137\n",
      "\n",
      "mRNA transcript variant 3 NM_001184740.2, 19 exons,  total annotated spliced exon length: 2255\n",
      "protein isoform c NP_001171669.1 (CCDS53676.1), 17 coding  exons,  annotated AA length: 634\n",
      "\n",
      "Exon table for  mRNA  NM_001184740.2 and protein NP_001171669.1\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t543\n",
      "70420400-70420510\t\t70420400-70420510\t\t21872-21982\t\t21872-21982\t\t111\t\t111\t\t2429\n",
      "70422940-70422995\t\t70422940-70422995\t\t24412-24467\t\t24412-24467\t\t56\t\t56\t\t2336\n",
      "70425332-70425401\t\t70425332-70425401\t\t26804-26873\t\t26804-26873\t\t70\t\t70\t\t3649\n",
      "70429051-70429199\t\t70429051-70429199\t\t30523-30671\t\t30523-30671\t\t149\t\t149\t\t1991\n",
      "70431191-70431280\t\t70431191-70431280\t\t32663-32752\t\t32663-32752\t\t90\t\t90\t\t1820\n",
      "70433101-70433278\t\t70433101-70433278\t\t34573-34750\t\t34573-34750\t\t178\t\t178\t\t368\n",
      "70433647-70433718\t\t70433647-70433718\t\t35119-35190\t\t35119-35190\t\t72\t\t72\t\t1307\n",
      "70435026-70435119\t\t70435026-70435119\t\t36498-36591\t\t36498-36591\t\t94\t\t94\t\t346\n",
      "70435466-70435744\t\t70435466-70435744\t\t36938-37216\t\t36938-37216\t\t279\t\t279\t\t537\n",
      "70436282-70436575\t\t70436282-70436408\t\t37754-38047\t\t37754-37880\t\t294\t\t127\n",
      "\n",
      "mRNA transcript variant X1 XM_006718447.4, 16 exons,  total annotated spliced exon length: 3027\n",
      "protein isoform X1 XP_006718510.1, 14 coding  exons,  annotated AA length: 476\n",
      "\n",
      "Exon table for  mRNA  XM_006718447.4 and protein XP_006718510.1\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t3083\n",
      "70422940-70422995\t\t70422940-70422995\t\t24412-24467\t\t24412-24467\t\t56\t\t56\t\t2336\n",
      "70425332-70425401\t\t70425332-70425401\t\t26804-26873\t\t26804-26873\t\t70\t\t70\t\t3649\n",
      "70429051-70429199\t\t70429051-70429199\t\t30523-30671\t\t30523-30671\t\t149\t\t149\t\t1991\n",
      "70431191-70431280\t\t70431191-70431280\t\t32663-32752\t\t32663-32752\t\t90\t\t90\t\t1820\n",
      "70433101-70433278\t\t70433101-70433278\t\t34573-34750\t\t34573-34750\t\t178\t\t178\t\t368\n",
      "70433647-70433718\t\t70433647-70433718\t\t35119-35190\t\t35119-35190\t\t72\t\t72\t\t1307\n",
      "70435026-70436575\t\t70435026-70435162\t\t36498-38047\t\t36498-36634\t\t1550\t\t137\n",
      "\n",
      "mRNA transcript variant X2 XM_006718448.5, 14 exons,  total annotated spliced exon length: 1601\n",
      "protein isoform X2 XP_006718511.1, 12 coding  exons,  annotated AA length: 369\n",
      "\n",
      "Exon table for  mRNA  XM_006718448.5 and protein XP_006718511.1\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t543\n",
      "70420400-70420510\t\t70420400-70420510\t\t21872-21982\t\t21872-21982\t\t111\t\t111\t\t959\n",
      "70421470-70421580\t\t70421470-70421580\t\t22942-23052\t\t22942-23052\t\t111\t\t111\t\t1359\n",
      "70422940-70422995\t\t70422940-70422995\t\t24412-24467\t\t24412-24467\t\t56\t\t56\t\t2336\n",
      "70425332-70425401\t\t70425332-70425401\t\t26804-26873\t\t26804-26873\t\t70\t\t70\t\t295\n",
      "70425697-70426087\t\t70425697-70425779\t\t27169-27559\t\t27169-27251\t\t391\t\t83\n",
      "\n",
      "mRNA transcript variant X3 XM_017017312.3, 11 exons,  total annotated spliced exon length: 1456\n",
      "protein isoform X3 XP_016872801.1, 9 coding  exons,  annotated AA length: 277\n",
      "\n",
      "Exon table for  mRNA  XM_017017312.3 and protein XP_016872801.1\n",
      "Genomic Interval Exon\t\tGenomic Interval Coding\t\tGene Interval Exon\t\tGene Interval Coding\t\tExon Length\tCoding Length\tIntron Length\n",
      "------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "70398529-70398614\t\t1-86\t\t86\t\t6650\n",
      "70405265-70405361\t\t6737-6833\t\t97\t\t1936\n",
      "70407298-70407384\t\t70407298-70407384\t\t8770-8856\t\t8770-8856\t\t87\t\t87\t\t133\n",
      "70407518-70407591\t\t70407518-70407591\t\t8990-9063\t\t8990-9063\t\t74\t\t74\t\t2239\n",
      "70409831-70409960\t\t70409831-70409960\t\t11303-11432\t\t11303-11432\t\t130\t\t130\t\t4581\n",
      "70414542-70414652\t\t70414542-70414652\t\t16014-16124\t\t16014-16124\t\t111\t\t111\t\t1010\n",
      "70415663-70415717\t\t70415663-70415717\t\t17135-17189\t\t17135-17189\t\t55\t\t55\t\t1295\n",
      "70417013-70417123\t\t70417013-70417123\t\t18485-18595\t\t18485-18595\t\t111\t\t111\t\t2622\n",
      "70419746-70419856\t\t70419746-70419856\t\t21218-21328\t\t21218-21328\t\t111\t\t111\t\t543\n",
      "70420400-70420510\t\t70420400-70420510\t\t21872-21982\t\t21872-21982\t\t111\t\t111\t\t1998\n",
      "70422509-70422991\t\t70422509-70422552\t\t23981-24463\t\t23981-24024\t\t483\t\t44\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "#\n",
    "# Utilisons Entrez.efetch\n",
    "#\n",
    "geneID = newLib[\"eSearchResult\"][\"IdList\"][\"Id\"]\n",
    "#\n",
    "# On crée le handle pour géer le flux de données\n",
    "#\n",
    "summary = Entrez.esummary(db=\"gene\", id=geneID)\n",
    "#\n",
    "# On crée un dictionnaire via la méthode read\n",
    "# pour récupérer les données obtenues\n",
    "#\n",
    "summary = Entrez.read(summary)\n",
    "#\n",
    "# Juste pour présenter ça beau ;-)\n",
    "# Essayer pour voir ;-)\n",
    "# Ça permet de voir la structure du dictionnaire\n",
    "#\n",
    "#summaryJ = json.dumps(summary,indent=2)\n",
    "#print(summaryJ)\n",
    "\n",
    "print(summary[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"Name\"])\n",
    "print(summary[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"Description\"])\n",
    "print(summary[\"DocumentSummarySet\"][\"DocumentSummary\"][0][\"Summary\"])\n",
    "\n",
    "#\n",
    "# Mais les données sur les transcrits et le protéines produites sont obtenues\n",
    "# autrement....\n",
    "#\n",
    "data = Entrez.efetch(db=\"gene\", id=geneID, rettype=\"gene_table\", retmode=\"text\")\n",
    "data = data.read()\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e581dc96-3bfa-406f-9c08-8b1b25039eb2",
   "metadata": {},
   "source": [
    "### Étape 4 - Obtenir les séquences pour chaque transcrit de la cortactine, 1ère partie\n",
    "\n",
    "Il nous faut obtenir les identificateurs des transcrits afin d'aller récupérér les séquences. On récupère le contenu de la variable data sous la forme d'un tableau de lignes créé grâce à la méthode splitlines(). Cette dernière brise une longue chaine de caractères (le contenu de data) en ligne en recherchant les caractères de bris de ligne comme \\n ou \\r.\n",
    "\n",
    "Après, il ne nous reste qu'à trier le tableau pour obtenir les infos nécessaires en récupérant l'information discriminante: NM_* ou XM_* (on pourrait faire de même pour les protéines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "10228638-4531-4dff-a9bf-6bd5fc858a59",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA transcript variant 1 NM_005231.4, 18 exons,  total annotated spliced exon length: 3249\n",
      "mRNA transcript variant 2 NM_138565.3, 17 exons,  total annotated spliced exon length: 3138\n",
      "mRNA transcript variant 3 NM_001184740.2, 19 exons,  total annotated spliced exon length: 2255\n",
      "mRNA transcript variant X1 XM_006718447.4, 16 exons,  total annotated spliced exon length: 3027\n",
      "mRNA transcript variant X2 XM_006718448.5, 14 exons,  total annotated spliced exon length: 1601\n",
      "mRNA transcript variant X3 XM_017017312.3, 11 exons,  total annotated spliced exon length: 1456\n",
      "['NM_005231.4', 'NM_138565.3', 'NM_001184740.2', 'XM_006718447.4', 'XM_006718448.5', 'XM_017017312.3']\n"
     ]
    }
   ],
   "source": [
    "allLines = data.splitlines()\n",
    "\n",
    "allTranscripts = []\n",
    "\n",
    "for aLine in allLines:\n",
    "    if \"mRNA transcript\" in aLine:\n",
    "        print(aLine)\n",
    "        tmp = aLine.split(\" \")\n",
    "        #\n",
    "        # Parce que l'on sait que le 5ème élément est\n",
    "        # l'identificateur unique pour Genbank\n",
    "        #\n",
    "        # Il faut retirer la virgule à la fin...\n",
    "        #\n",
    "        allTranscripts.append(tmp[4].replace(',',''))\n",
    "\n",
    "print(allTranscripts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83f5ede3-30b6-4b83-9d55-5d7c1f47905a",
   "metadata": {},
   "source": [
    "### Étape 5 - Obtenir les séquences pour chaque transcrit de la cortactine, 2ème partie\n",
    "\n",
    "Ok, on a un tableau pour les identificateurs des séquences de chaque transcrit! Allons maintenant récupérer les infos avec `Entrez.efetch`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3eed2a74-ad85-49c9-81d6-e742df9c9d37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: NM_138565.3\n",
      "Name: NM_138565\n",
      "Description: Homo sapiens cortactin (CTTN), transcript variant 2, mRNA\n",
      "Number of features: 23\n",
      "/molecule_type=mRNA\n",
      "/topology=linear\n",
      "/data_file_division=PRI\n",
      "/date=07-APR-2024\n",
      "/accessions=['NM_138565']\n",
      "/sequence_version=3\n",
      "/keywords=['RefSeq']\n",
      "/source=Homo sapiens (human)\n",
      "/organism=Homo sapiens\n",
      "/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n",
      "/references=[Reference(title='METTL3 promotes osteogenic differentiation of human umbilical cord mesenchymal stem cells by up-regulating m6A modification of circCTTN', ...), Reference(title='A MAP1B-cortactin-Tks5 axis regulates TNBC invasion and tumorigenesis', ...), Reference(title='Requirement of Site-Specific Tyrosine Phosphorylation of Cortactin in Retinal Neovascularization and Vascular Leakage', ...), Reference(title='Binding of USP4 to cortactin enhances cell migration in HCT116 human colon cancer cells', ...), Reference(title='Cortactin expression in a Norwegian cohort of human papilloma virus negative oral squamous cell carcinomas of the mobile tongue', ...), Reference(title='Initial assessment of human gene diversity and expression patterns based upon 83 million nucleotides of cDNA sequence', ...), Reference(title='The product of the EMS1 gene, amplified and overexpressed in human carcinomas, is homologous to a v-src substrate and is located in cell-substratum contact sites', ...), Reference(title='Amplified region of chromosome band 11q13 in breast and squamous cell carcinomas encompasses three CpG islands telomeric of FGF3, including the expressed gene EMS1', ...), Reference(title='Cortactin, an 80/85-kilodalton pp60src substrate, is a filamentous actin-binding protein enriched in the cell cortex', ...), Reference(title='Identification and cloning of two overexpressed genes, U21B31/PRAD1 and EMS1, within the amplified chromosome 11q13 region in human carcinomas', ...)]\n",
      "/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "reference sequence was derived from DA716772.1, BC008799.2 and\n",
      "AK222613.1.\n",
      "On Aug 14, 2020 this sequence version replaced NM_138565.2.\n",
      "Summary: This gene is overexpressed in breast cancer and squamous\n",
      "cell carcinomas of the head and neck. The encoded protein is\n",
      "localized in the cytoplasm and in areas of the cell-substratum\n",
      "contacts. This gene has two roles: (1) regulating the interactions\n",
      "between components of adherens-type junctions and (2) organizing\n",
      "the cytoskeleton and cell adhesion structures of epithelia and\n",
      "carcinoma cells. During apoptosis, the encoded protein is degraded\n",
      "in a caspase-dependent manner. The aberrant regulation of this gene\n",
      "contributes to tumor cell invasion and metastasis. Three splice\n",
      "variants that encode different isoforms have been identified for\n",
      "this gene. [provided by RefSeq, May 2010].\n",
      "Transcript Variant: This variant (2) lacks an internal segment of\n",
      "sequence, compared to variant 1. The resulting protein (isoform b)\n",
      "has the same N- and C- termini but is shorter when it is compared\n",
      "to isoform a.\n",
      "Publication Note:  This RefSeq record includes a subset of the\n",
      "publications that are available for this gene. Please see the Gene\n",
      "record to access additional publications.\n",
      "/structured_comment=defaultdict(<class 'dict'>, {'Evidence-Data': {'Transcript exon combination': 'BC008799.2, SRR18074968.3168512.1 [ECO:0000332]', 'RNAseq introns': 'single sample supports all introns SAMEA1965299, SAMEA1966682 [ECO:0000348]'}})\n",
      "Seq('GGAACCGGAAGTAGAGCCTGGTGCCTGGGAGCGGCTGGCGCGGCGGAATCCAGG...CCA')\n",
      "GGAACCGGAAGTAGAGCCTGGTGCCTGGGAGCGGCTGGCGCGGCGGAATCCAGGGCCGACCCGGGCCGGACCGACCCCAGGCGGCGACGGAATCAGTCCCCAATGCCTGGAAATTCCTCATTGGATTACTGTGTTTTAAACAGAATTTCGTGAACAGCCTTTTATCTCCAAGCGGAAAGAAAGATGTGGAAAGCTTCAGCAGGCCACGCTGTGTCCATCGCCCAGGATGACGCGGGGGCCGATGACTGGGAGACCGACCCTGATTTTGTGAATGATGTGAGTGAGAAGGAGCAAAGATGGGGTGCCAAGACGGTGCAGGGCTCCGGGCACCAGGAGCATATCAACATACACAAGCTGAGGGAGAATGTCTTTCAAGAGCATCAGACCCTTAAGGAGAAGGAACTTGAAACAGGACCAAAAGCTTCCCATGGCTATGGAGGGAAATTTGGTGTGGAACAAGACCGAATGGATAAGTCAGCTGTCGGCCACGAATATCAGTCGAAACTTTCCAAGCACTGCTCGCAGGTGGACTCGGTCCGTGGCTTCGGAGGCAAGTTTGGTGTCCAGATGGACAGAGTTGATCAGTCTGCTGTAGGCTTTGAATACCAGGGGAAGACTGAGAAGCATGCCTCCCAGAAAGACTACTCCAGTGGTTTTGGCGGCAAGTATGGCGTGCAGGCCGACCGAGTAGACAAGAGCGCGGTGGGCTTCGACTACCAGGGCAAGACGGAGAAGCACGAGTCACAGAGAGATTACTCCAAAGGTTTCGGCGGCAAATACGGTATCGACAAGGACAAAGTGGATAAGAGCGCCGTTGGCTTTGAGTATCAAGGCAAAACGGAGAAGCACGAGTCCCAGAAAGACTATGTGAAAGGGTTTGGAGGAAAATTTGGTGTGCAGACAGACAGACAAGACAAATGTGCCCTTGGCTGGGATCACCAGGAGAAATTGCAGCTGCATGAATCCCAAAAAGACTACTCCAAAGGATTCGGCGGGAAGTATGGGGTGCAGAAGGATCGGATGGATAAGAATGCGTCAACCTTTGAGGATGTCACCCAGGTGTCCTCTGCCTACCAGAAGACAGTACCTGTCGAAGCTGTGACCAGCAAAACAAGTAACATCAGAGCTAACTTTGAAAACCTCGCTAAGGAGAAAGAGCAGGAGGACAGGCGGAAGGCGGAGGCGGAGAGAGCCCAGCGGATGGCCAAGGAGCGGCAGGAGCAGGAAGAGGCCAGGAGGAAGCTGGAGGAGCAAGCCAGAGCCAAAACGCAAACGCCCCCTGTGTCGCCCGCACCTCAGCCAACCGAGGAGAGGCTGCCCTCGAGCCCCGTCTATGAGGATGCGGCTTCCTTCAAGGCAGAGCTGAGCTACAGAGGCCCTGTGAGTGGGACGGAGCCGGAGCCCGTGTACAGCATGGAGGCCGCTGACTACCGAGAGGCCAGCAGCCAGCAGGGCCTGGCCTATGCCACAGAGGCTGTCTATGAAAGCGCAGAGGCCCCGGGCCACTATCCCGCAGAGGACAGCACCTACGATGAGTACGAGAACGATCTGGGGATCACAGCCGTCGCCCTGTACGACTACCAGGCTGCGGGCGATGATGAGATCTCATTTGACCCTGATGACATCATCACCAACATCGAGATGATTGACGACGGCTGGTGGCGCGGGGTGTGCAAGGGCCGGTACGGGCTCTTCCCAGCCAACTATGTGGAGCTGCGGCAGTAGGGCCCCCAGCCCCCCCCCGGAGCTGCGCCCTGGATCCTCACACTACAGATCAGGCCTTCTTTGGTTCTTGGGTGGTTTTGGGTTTTTTCTGTTTTTTTTTTTTTTTTTTTTTTTTTGAAGGTGGGGAGGGGAATATACACATTGCTTTTATATTTAATACTTTTGCTGATGCTTTTGAAAATGTTTATGCCACAGAATTTGCTAATATATTGTAATCACATTCCTTAGGAGGACTTTGGTAATTGGTTTTATGCATTGATGGTTTTTTTTTTCTTTTTTGCCAAATTGACTGTCACGCGGCAGCTTCAGGGAGCTCGCATTCTCTTGTGTTCGTGTTGCCCTCGTGCCCATCAAGTGCAGTCGGGACCTCCCAGGACAAGCACGAGGCCTCAGGTCGGCCCTGTGGCGGGTAGGCAGGAAGGACTGTCCCAGACGAGGGGCTTCCTCTAGAGTCTCACTGCTGGGGAGGAGAGGACTGGGCCTGATGGAAGTTAACCCGGAGCTAAGTCACCCAGAGCACAGGAGCTGCCATGTCAGATGGGAAATCTGCCTATGTCATACCGTGACAGCCCGCAGGATCAGGTGACTTCTAGCAGAGACCCTGGTTTTTTTCCTGTGCCCACTCCGGCTTGTCCTCATCTCTACCCATCCCCTGATGCCCAGGTCACCGGGAGGGCTGCTGGGAGCCTCTCCTGTCCCCGCCGGGCAGTGTCACTGAGTCCTTGAAATCCTCCCCTGCCCCGCGGGTCTCTGGATTGGGACGCACAGTGCAGTTGAGGTCTGCGTCGGGCTTGGCTTTTCACAAAGGCTGATGTCTTAACTGTCACCCATATGGTCCCTGGGCCACCGGGCAGCCTGGGGCGGTGTGTGTGCCATGTCACAGCATGGCCTCTCGGCCTTGGGAAGGAAGGCAGTGCCTGCTCTGCTGTGAGCCGCCAGGAACCCTCCTCCTGTCAATGGGGGTGTAGTATTTTTGCCAAAATATCATGTTCAATTTCAGTAGTTTGATCAGTTGAAGGCTAGAAGTGTGAAGTGCAGATGAGTGTGTGTTCTTCCCCAAGGTCCCCCCACAGCTCCAGGACACCGCTGTCCTGGCATTTGTGGCCACTCACTTTGTAGGAAACTCATCTCCTTCCTGAGGAGCCGGGAGGCTGGACCAGTCCCGTCGTGCAGTCAGGTGGGCGGTGTGTCTTTCCAGAAGGTCACGTGGAAATGTCTCGGGACTTGGGTCCCGGAGTGCCCGTGAAGCGTGTTTTTGCTCCTGAGGTGCATTTTCTCATCATCCTTGCTTTACCACAATGAGCAATGAGGTCGGGTTTTATATGCAACTTATTGTATCTGAATTCCTGTAGCACACCTCATAGGTATGATTTTTTTAAATTAAAGAATTCAGAATAAACATTTTTTGATCCA\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# Nous allons procéder via une boucle for en itérant sur les items de la liste allTranscripts\n",
    "# pour ne garder que ceux qui commencent par NM_\n",
    "#\n",
    "# Créons une lsite vide d'objets Entrez.Seq\n",
    "#\n",
    "\n",
    "mySequences = []\n",
    "\n",
    "for i in allTranscripts:\n",
    "    if \"NM_\" in i:\n",
    "        fluxe = Entrez.efetch(db=\"nucleotide\", id=i, rettype=\"gb\", retmode=\"text\")\n",
    "        #\n",
    "        # On se doit de transformer le flux de données en objet Entrez.Seq\n",
    "        #\n",
    "        aSeq = SeqIO.read(fluxe,\"gb\")\n",
    "        mySequences.append(aSeq)\n",
    "#\n",
    "# Regardons uniquement les infos du 2ème transcrit\n",
    "#\n",
    "print(mySequences[1])\n",
    "print(mySequences[1].seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b884a56-672e-4fe0-9af7-ef241a47bc60",
   "metadata": {},
   "source": [
    "### Étape 6 - Écrire nos résultats dans des fichiers\n",
    "\n",
    "Utilisons maintenant `SeqIO` pour écrire nos résultats dans des fichiers, un par objet `Seq`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "50325d84-9e4e-458a-8f1a-5e5963b85ff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "# SeqIO permet non seulement la lecture mais aussi l'écriture\n",
    "# de fichiers dans divers formats\n",
    "#\n",
    "for i in mySequences:\n",
    "    #\n",
    "    # Profitons des méthodes de l'objet Seq pour déterminer le nom\n",
    "    # du fichier\n",
    "    #\n",
    "    # Gardons notre fichier simple en utilisant le format FASTA\n",
    "    #\n",
    "    SeqIO.write(i,\"../z.misc_files/data_seq/\"+i.name+\".fa\",\"fasta\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
